### PR TITLE
BDOG-1212 Support release candidates

### PR DESCRIPTION
**Pull request by colin-lamed**

### CODE REVIEW

File: src/main/scala/uk/gov/hmrc/versioning/ReleaseVersioning.scala [hunk 1]
```diff
@@ -21,58 +21,86 @@ object ReleaseVersioning {
...
 
-  private val tag = """^(?:release\/|v)?(\d+)\.(\d+)\.(\d+)(?:-\d+-g[a-z0-9]{4,40}$)?""".r
+      case _ =>
+        throw new IllegalArgumentException(s"invalid version format for '$v'")
+    }
+  }
 
-  private object AsInt {
-    def unapply(arg: String): Option[Int] = Some(arg.toInt)
+  case class Version(
```
Comment (chotai) [inline src/main/scala/uk/gov/hmrc/versioning/ReleaseVersioning.scala]: should we make this private?

**Review by christopherjturner — DISMISSED** [review]

File: build.sbt [hunk 2]
```diff
@@ -1,24 +1,16 @@
...
   .settings(defaultSettings(): _*)
   .settings(
-    scalaVersion := "2.11.12",
+    scalaVersion := "2.12.14",
     crossScalaVersions := Vector("2.10.7", "2.11.12", "2.12.8"),
     majorVersion := 0,
-    makePublicallyAvailableOnBintray := true,
+    isPublicArtefact := true,
     libraryDependencies ++= compileDependencies ++ testDependencies,
     retrieveManaged := true,
```
Comment (chotai) [inline build.sbt]: Do we need this?

File: src/main/scala/uk/gov/hmrc/versioning/ReleaseVersioning.scala [hunk 3]
```diff
@@ -21,58 +21,86 @@ object ReleaseVersioning {
...
 
-    case Some(tag(AsInt(major), _, _)) if requestedMajorVersion != major =>
-      s"$requestedMajorVersion.0.0"
+      // was previously a releaseCandidate, but not anymore
+      case Some(Version(major, minor, patch, Some(_))) if !releaseCandidate =>
+        s"$major.$minor.$patch"
 
-    case Some(tag(major, minor, AsInt(patch))) if hotfix =>
-      s"$major.$minor.${patch + 1}"
+      case Some(Version(major, _, _, optRc)) if requestedMajorVersion != major =>
```
Comment (chotai) [inline src/main/scala/uk/gov/hmrc/versioning/ReleaseVersioning.scala]: can we make the order of `requestedMajorVersion != major` consistent between this and line 46 please?

**Review by chotai — DISMISSED** [review]

File: build.sbt (hunk 2 above)
Comment (colin-lamed) [reply build.sbt]: Probably not. Looks like it was broken with coursier anyway (https://github.com/sbt/sbt/issues/5078)

**Review by chotai — APPROVED** [review]

**Pull request merged**